core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -215,45 +242,67 @@ public void runInContext() {
     if (compressor != Codec.Identity.NONE) {
       stream.setMessageCompression(true);
     }
+
     // Delay any sources of cancellation after start(), because most of the transports are broken if
     // they receive cancel before start. Issue #1343 has more details
 
-    // Start the deadline timer after stream creation because it will close the stream
-    if (callOptions.getDeadline() != null) {
-      long timeoutNanos = callOptions.getDeadline().timeRemaining(NANOSECONDS);
-      deadlineCancellationFuture = startDeadlineTimer(timeoutNanos);
-      if (deadlineCancellationFutureShouldBeCancelled) {
-        // Race detected! ClientStreamListener.closed may have been called before
-        // deadlineCancellationFuture was set, thereby preventing the future from being cancelled.
-        // Go ahead and cancel again, just to be sure it was cancelled.
-        deadlineCancellationFuture.cancel(false);
-      }
-    }
     // Propagate later Context cancellation to the remote side.
-    this.context.addListener(this, directExecutor());
+    context.addListener(this, directExecutor());
+    if (contextListenerShouldBeRemoved) {
+      // Race detected! ClientStreamListener.closed may have been called before
+      // deadlineCancellationFuture was set, thereby preventing the future from being cancelled.
+      // Go ahead and cancel again, just to be sure it was cancelled.
+      context.removeListener(this);
+    }
   }
 
   /**
    * Based on the deadline, calculate and set the timeout to the given headers.
-   *
-   * @return {@code false} if deadline already exceeded
    */
-  static boolean updateTimeoutHeader(@Nullable Deadline deadline, Metadata headers) {
-    // Fill out timeout on the headers
-    // TODO(carl-mastrangelo): Find out if this should always remove the timeout,
-    // even when returning false.
+  private static void updateTimeoutHeaders(@Nullable Deadline effectiveDeadline,
+      @Nullable Deadline callDeadline, @Nullable Deadline outerCallDeadline, Metadata headers) {
+    // TODO: Ask Louis: What if timeout is not specified in CallOptions but set in metadata?
     headers.removeAll(TIMEOUT_KEY);
```

Comment:
Generally we've taken the approach of always removing such metadata (independent of whether deadline is set elsewhere). So the removal here seems a Good Thing™ to me so that users are forced to use the correct API.